### ISSUE TITLE
Crash when recording in a new unsaved project for a longer period of time

### ISSUE THREAD

**Issue by pasisavolainen** [audacity/audacity#7018]
### Stacktrace

```
Operating system: Windows NT
                  10.0.22631 3958
CPU: amd64
     family 25 model 97 stepping 2
     32 CPUs

GPU: UNKNOWN // nvidia 3090 if that matters

Crash reason:  EXCEPTION_ACCESS_VIOLATION_READ
Crash address: 0x0
Process uptime: 4889 seconds

Thread 0 (crashed)
 0  lib-wave-track.dll + 0xf16b
... // rest attached as file
```

### Steps to reproduce

1. launch Audacity
2. set recording channels to 2
3. click Record
4. wait for ~1-1.5h

**UPDATE by PS**
<details>
Another simple reproducible use case _(see first comment below in this thread)_
1. clear audacity settings folder (to get OOTB factory default settings)
5. launch 3.6.1 - on W10 or W11
6. Press Record
7. Observe: stereo recording starts
8. Observe: after around 6 hours Audacity crashes and stops recording

_Note that accepting recovery on relaunch safely restores the 6 hours of recorded audio._

**UPDATE by PS**
Note further that when saving an empty project before starting to record, then the crash does not occur.  Indiacting that the underlying issue here is with thu use of the temporary files.  I have thus edited the bug's title.</details>

### Expected behavior

no crash

### Actual behavior

a crash

### Audacity Version

3.6.0 onwards

### Operating system

Windows

### Additional context

Recording from Xonar U7 MKIII Line-In (USB soundbox by asus), MOTU M2 interface.

**UPDATE by PS**
This is a regression on 3.5.1

**Comment by petersampsonaudacity** [comment]
I can confirm getting crashes with long recordings on Windws, both W10 1nd W11 - in both cases Audacity crashes and stops stereo recording at around 6 hours - and in both cases re-launch properly recovers the c. 6 hours of recorded audio.

The basic test is to use Audacity 3.6.1 in its default settings and just set it to record in stereo.

### 1) Windows 10
In the course of testing  **Crash when closing or exiting Audacity with large unsaved project** #6890 
I ran a long recording test on my W10 laptop in Manchester.  This PC has a 256GB SSD and a 1TB onboard HDD - the temp files and thus the recordingwere all on the fast SSD.

It crashed during recording at a little after 6 hours.

There is plenty of disk space, the Audacity Status Bar says 28 hours available for recording.

### 2) Windows 11
Now in Zürich, my W11 PC here just has a 256GB SSD.

This crashed after 5h 56m 22s

Here too there is plenty of free disk space - the Audacity Status bar at start of recoe=rding shpws space for 84 hours of recordng.

**Comment by petersampsonaudacity** [comment]
@pasisavolainen 

The failure to send the crash report is a known and logged bug that currently remains unfixed (I think Muse do not know what causes this).

See:   **Crash report fails to send** #4154

**Comment by petersampsonaudacity** [comment]
@chinakov 

Hi Anton,  further testing on my W11laptop seems to point the finger at the use of the temporary files in an unsaved project as the potential culprit for this crash.  

I'm suspecting that somewhre there is a "hidden" limit to the maximum siz of the temporay files folder sufficient for 6 hours of stereo or 12 hours of mono at default settings

Yesterday and overnight I made a 17 hour stereo recording (with no crash at ~6 hours)

STRs
1. purge Audacity settings folder _(to get OOTB factory default settings)_
2. launch 3.6.1
3. Save the empty project _(in my case on my fast 256GB SSD)_
4. Press Record
5. Observe: Status bar shows 84 hours available for recording
6. Wait for many hours _(certainly greater than 6-7 hours)_
7. 17 hours later press Stop
8. Observe: 17 hours of recording
9. Press Exit or Quit
10. Observe:  there is **NO** crash
11. Relaunch Audacity
12. Observe: no recovery offered
13. Open the project that was saved at step 2
14. Observe: empty project _(expected behavior)_

---------------------------------------------------------------------------------

It is unclear from @pasisavolainen 's Steps To Reproduce whether his second project is pre-saved or just using the temporary files, but it looks ti me as though it is likely that his 4th step does not involve a Save Project before recording and is thus using the temporary files.

**Comment by pasisavolainen** [comment]
I was not pre-saving project files, just starting new instance of Audacity and pressing record.

Ultimately also "first" instance started crashing after a short while (5min recording) but then when restarted it recorded 30min without issues

**Comment by petersampsonaudacity** [comment]
@chinakov 

Hi Anton, 

this is actually a regression, but on which precise version I am unclear at the moment.

But certainly testesting with 2.4.2 yesterday/overnight I recorded 17h 50m of stereo with no crash and no recovery on relaunch.  

It will be hard (tedious) to ascertain precisely in which version the regression occurred as each tests involves many elapsed hours. Right now I am testing with 3.0.5 (the last version produced by the old Audacity Team).

**Comment by petersampsonaudacity** [comment]
@chinakov @dozzzzer 

Something wnnt seriosly wrong with recording using temp files (un-saved project) between 3.5/0 and 3.6.0.

3.5.- recorded fine for me yesterday with temp files for 10h 11m

But this morning when I tried with 3.6.0 it vrashed after a short time on each try.  The first was 2m 40s, the second was a mere 17.83 seconds and the third try managed 13m 58s.  Here is a screenshot of the 3rd failure.
![image](https://github.com/user-attachments/assets/e17996cd-d996-4f7a-bfa6-e43c2ac20af1)

===================================================================

Right now I have 3,5.1 running a test to narrow down the precise version where the regression occurred, but the results of that are likely to be 10 plus hours away.

_I will also re-run 3.6.1 and later 3.6.2 Beta When I get time) but you will recall that earlier on W10 and W11 3,6.1 ran for around 6 hours beforer crashing._

**Comment by petersampsonaudacity** [comment]
@chinakov @dozzzzer 

Well now I'm totally baffled ...

I set a recording running yesterday evening using 3.6.0 expecting it to crash out - BUT instead this norning it was still happily running at 11h 05m when I stopped it.  So this looks like a "moonphase" bug.

_Right now I have the latest 3.6.2 Beta build running a stereo recording.  Currently it is happily recording away at 1h 05m_
_So time will tell, later on today ..._

**Comment by dozzzzer** [comment]
@petersampsonaudacity thanks for the report. Looking into it now.

**Comment by dozzzzer** [comment]
Yes, I can reproduce that on my end too. Thanks again, Peter, we will definitely prioritize this for 3.6.2.

**Comment by dozzzzer** [comment]
UPD: after 2+ hrs of recording on MacOS, I didn't manage to reproduce the crash.

**Comment by petersampsonaudacity** [comment]
@dozzer  

note that earlier in this thread I need circa 6 hours of recording on both W10 and W11 to get the crash

**Comment by dozzzzer** [comment]
Noted 👍

**Comment by petersampsonaudacity** [comment]
@dozzzzer 

Sergey:  Re your changed bug title ...

I added the "Whenusing temporary files" earkier as I found I could not reproduce the carsh if I saved an empty project before recording, thereby avoiding the use of temporary files.  Which leads ne to guess that this may be a temp files issue

**Comment by dozzzzer** [comment]
@kryksyh the issue doesn't seem to be present in 3.5.1

**Comment by kryksyh** [comment]
@dozzzzer thanks for the heads-up. Do you have any hints on how I can reproduce it? I tried to speed things up by setting a very high sample rate, but so far no crashes after an hour and a half in.

**Comment by dozzzzer** [comment]
I'm afraid, I don't know how to speed this up 😞 Normal recording in 44.1 kHz/stereo does the job for me after about 45-60 minutes.

**Comment by dozzzzer** [comment]
Tested

**Comment by petersampsonaudacity** [comment]
Testing on W11 with @kryksyh 's latest branch build for this:  audacity-win-3.6.2-alpha-20240814+4074e72-x64

I got a crash at 2h 7m 19s
![image](https://github.com/user-attachments/assets/9129e424-f1ac-47ab-b500-b386837ed8a7)

**Comment by petersampsonaudacity** [comment]
We are getting a **LOT** of complaints on the Forum about crashes - many users are being affected by this.

**Comment by dozzzzer** [comment]
I just got a crash too after 5h 43m of recording.

**Comment by kryksyh** [comment]
@petersampsonaudacity @dozzzzer thanks. I also got a crash, and from the dump file it generated the reason is different from the fix in this PR. So it seems that we have at least two different bugs resulting in a crash during recording.

**Comment by petersampsonaudacity** [comment]
Testing on W11 with @kryksyh 's latest branch test buid for this:  audacity-win-3.6.2-alpha-20240820+63fc474-x64

Hi Dmitry,

last chance to test on W11 for a while - and the most I can let it run for is an hour or so - as we are packing up in Zurich, returning to Manchester and W10 later today.

### UPDATE
Ih 41m layer still recording nicely.  Stopped the recording

**Comment by petersampsonaudacity** [comment]
Testing on W10 with @kryksyh 's latest branch test build for this:  audacity-win-3.6.2-alpha-20240820+63fc474-x64

Hi Dmitry,

This latest test build had the recording running for 16 hours 30 minutes - with no crash.

_This chewed up most of the remaining free space on my SSD C: drive, which thankfully got released properly once I exited the project without saving._

**Comment by dozzzzer** [comment]
While testing this PR's build on Windows 10, I encountered a freeze after some 12.5 hrs of recording.

**Comment by petersampsonaudacity** [comment]
@kryksyh 

Hi Dmitry,

so now I am testing the very latest £.6.2 Beta build using my slower 1TB HD as the location for the temporary files (the app of course remains on my SSD C: drive)

Will report back later.

**Comment by dozzzzer** [comment]
22+ hrs of recording and no crash (freeze) so far.

**Comment by petersampsonaudacity** [comment]
Ok so the test with the temp files on my USB stick is still rumming sweetly at

**Comment by petersampsonaudacity** [comment]
@dozzzzer 

Can this be closed now that @kryksyh 's fix has been pulled into 3.6.2 ?

_Or are you leaving it open just in case ..._
